### ISSUE TITLE
[OrionMySQLSink] Cygnus does not reconnect to MySQL server (HA tests)

### ISSUE THREAD

**Issue by manucarrace** [telefonicaid/fiware-cygnus#734]
When mysql process is restarted the cygnus is not able to reconnect. It will fail for the following requests untill cygnus process is restarted

**Comment by frbattid** [comment]
Cygnus opens a connection per each database, and each time a query is going to be executed in a database, this method is called:

```
        public Connection getConnection(String dbName) throws Exception {
            try {
                // FIXME: the number of cached connections should be limited to a certain number; with such a limit
                //        number, if a new connection is needed, the oldest one is closed
                Connection con = connections.get(dbName);

                if (con == null || !con.isValid(0)) {
                    if (con != null) {
                        con.close();
                    } // if

                    con = createConnection(dbName);
                    connections.put(dbName, con);
                } // if

                return con;
            } catch (ClassNotFoundException e) {
                throw new CygnusPersistenceError(e.getMessage());
            } catch (SQLException e) {
                throw new CygnusPersistenceError(e.getMessage());
            } // try catch
        } // getConnection
```

As can be seen, each time it is invoked (i.e. each time a query is going to be executed) the existence of the connection is checked, and if exists, it s checked as valir or not.

Regarding the `isValid()` method, this is what the documentation says:

> Returns true if the connection has not been closed and is still valid. The driver shall submit a query on the connection or use some other mechanism that positively verifies the connection is still valid when this method is called.
>
> The query submitted by the driver to validate the connection shall be executed in the context of the current transaction.

Such a method says a testing query is sent behind the scenes. That's the reason it is preferred instead of `isClosed()` because it only returns "closed" after effectively closing the connection by the client:

> Retrieves whether this Connection object has been closed. A connection is closed if the method close has been called on it or if certain fatal errors have occurred. This method is guaranteed to return true only when it is called after the method Connection.close has been called.
>
> This method generally cannot be called to determine whether a connection to a database is valid or invalid. A typical client can determine that a connection is invalid by catching any exceptions that might be thrown when an operation is attempted.

So, everything seems to be OK...

**Comment by frbattid** [comment]
I've tested the branch `release/0.11.0` with a continuous "stream" of notifications; 1 notification per second; temperature and pressure attributes regarding a typical Room1 entity; each second those attributes are incremeted by 1.

The test is about persisting the first notifications, then stop the MySQL server, wait for a while and then restart the server in order to check if Cygnus reconnects.

This is the content of the created table:

```
mysql> select * from reconnect_room1_room;
+---------------+--------------------------+-------------------+----------+------------+-------------+------------+-----------+----------------------------------------------------------+
| recvTimeTs    | recvTime                 | fiwareservicepath | entityId | entityType | attrName    | attrType   | attrValue | attrMd                                                   |
+---------------+--------------------------+-------------------+----------+------------+-------------+------------+-----------+----------------------------------------------------------+
| 1455261139603 | 2016-02-12T07:12:19.603Z | reconnect         | Room1    | Room       | temperature | centigrade | 0         | [{"name":"resolution","type":"integer","value":"0"}]     |
| 1455261139603 | 2016-02-12T07:12:19.603Z | reconnect         | Room1    | Room       | pressure    | pascal     | 0         | [{"name":"resolution","type":"integer","value":"0"}]     |
| 1455261140697 | 2016-02-12T07:12:20.697Z | reconnect         | Room1    | Room       | temperature | centigrade | 1         | [{"name":"resolution","type":"integer","value":".50"}]   |
| 1455261140697 | 2016-02-12T07:12:20.697Z | reconnect         | Room1    | Room       | pressure    | pascal     | 1         | [{"name":"resolution","type":"integer","value":".50"}]   |
| 1455261141728 | 2016-02-12T07:12:21.728Z | reconnect         | Room1    | Room       | temperature | centigrade | 2         | [{"name":"resolution","type":"integer","value":"1.00"}]  |
| 1455261141728 | 2016-02-12T07:12:21.728Z | reconnect         | Room1    | Room       | pressure    | pascal     | 2         | [{"name":"resolution","type":"integer","value":"1.00"}]  |
| 1455261142755 | 2016-02-12T07:12:22.755Z | reconnect         | Room1    | Room       | temperature | centigrade | 3         | [{"name":"resolution","type":"integer","value":"1.50"}]  |
| 1455261142755 | 2016-02-12T07:12:22.755Z | reconnect         | Room1    | Room       | pressure    | pascal     | 3         | [{"name":"resolution","type":"integer","value":"1.50"}]  |
| 1455261143784 | 2016-02-12T07:12:23.784Z | reconnect         | Room1    | Room       | temperature | centigrade | 4         | [{"name":"resolution","type":"integer","value":"2.00"}]  |
| 1455261143784 | 2016-02-12T07:12:23.784Z | reconnect         | Room1    | Room       | pressure    | pascal     | 4         | [{"name":"resolution","type":"integer","value":"2.00"}]  |
| 1455261144816 | 2016-02-12T07:12:24.816Z | reconnect         | Room1    | Room       | temperature | centigrade | 5         | [{"name":"resolution","type":"integer","value":"2.50"}]  |
| 1455261144816 | 2016-02-12T07:12:24.816Z | reconnect         | Room1    | Room       | pressure    | pascal     | 5         | [{"name":"resolution","type":"integer","value":"2.50"}]  |
| 1455261145850 | 2016-02-12T07:12:25.850Z | reconnect         | Room1    | Room       | temperature | centigrade | 6         | [{"name":"resolution","type":"integer","value":"3.00"}]  |
| 1455261145850 | 2016-02-12T07:12:25.850Z | reconnect         | Room1    | Room       | pressure    | pascal     | 6         | [{"name":"resolution","type":"integer","value":"3.00"}]  |
| 1455261146882 | 2016-02-12T07:12:26.882Z | reconnect         | Room1    | Room       | temperature | centigrade | 7         | [{"name":"resolution","type":"integer","value":"3.50"}]  |
| 1455261146882 | 2016-02-12T07:12:26.882Z | reconnect         | Room1    | Room       | pressure    | pascal     | 7         | [{"name":"resolution","type":"integer","value":"3.50"}]  |
| 1455261147921 | 2016-02-12T07:12:27.921Z | reconnect         | Room1    | Room       | temperature | centigrade | 8         | [{"name":"resolution","type":"integer","value":"4.00"}]  |
| 1455261147921 | 2016-02-12T07:12:27.921Z | reconnect         | Room1    | Room       | pressure    | pascal     | 8         | [{"name":"resolution","type":"integer","value":"4.00"}]  |
| 1455261148954 | 2016-02-12T07:12:28.954Z | reconnect         | Room1    | Room       | temperature | centigrade | 9         | [{"name":"resolution","type":"integer","value":"4.50"}]  |
| 1455261148954 | 2016-02-12T07:12:28.954Z | reconnect         | Room1    | Room       | pressure    | pascal     | 9         | [{"name":"resolution","type":"integer","value":"4.50"}]  |
| 1455261149981 | 2016-02-12T07:12:29.981Z | reconnect         | Room1    | Room       | temperature | centigrade | 10        | [{"name":"resolution","type":"integer","value":"5.00"}]  |
| 1455261149981 | 2016-02-12T07:12:29.981Z | reconnect         | Room1    | Room       | pressure    | pascal     | 10        | [{"name":"resolution","type":"integer","value":"5.00"}]  |
| 1455261151011 | 2016-02-12T07:12:31.11Z  | reconnect         | Room1    | Room       | temperature | centigrade | 11        | [{"name":"resolution","type":"integer","value":"5.50"}]  |
| 1455261151011 | 2016-02-12T07:12:31.11Z  | reconnect         | Room1    | Room       | pressure    | pascal     | 11        | [{"name":"resolution","type":"integer","value":"5.50"}]  |
| 1455261152074 | 2016-02-12T07:12:32.74Z  | reconnect         | Room1    | Room       | temperature | centigrade | 12        | [{"name":"resolution","type":"integer","value":"6.00"}]  |
| 1455261152074 | 2016-02-12T07:12:32.74Z  | reconnect         | Room1    | Room       | pressure    | pascal     | 12        | [{"name":"resolution","type":"integer","value":"6.00"}]  |
| 1455261153106 | 2016-02-12T07:12:33.106Z | reconnect         | Room1    | Room       | temperature | centigrade | 13        | [{"name":"resolution","type":"integer","value":"6.50"}]  |
| 1455261153106 | 2016-02-12T07:12:33.106Z | reconnect         | Room1    | Room       | pressure    | pascal     | 13        | [{"name":"resolution","type":"integer","value":"6.50"}]  |
| 1455261154145 | 2016-02-12T07:12:34.145Z | reconnect         | Room1    | Room       | temperature | centigrade | 14        | [{"name":"resolution","type":"integer","value":"7.00"}]  |
| 1455261154145 | 2016-02-12T07:12:34.145Z | reconnect         | Room1    | Room       | pressure    | pascal     | 14        | [{"name":"resolution","type":"integer","value":"7.00"}]  |
| 1455261155176 | 2016-02-12T07:12:35.176Z | reconnect         | Room1    | Room       | temperature | centigrade | 15        | [{"name":"resolution","type":"integer","value":"7.50"}]  |
| 1455261155176 | 2016-02-12T07:12:35.176Z | reconnect         | Room1    | Room       | pressure    | pascal     | 15        | [{"name":"resolution","type":"integer","value":"7.50"}]  |
| 1455261156209 | 2016-02-12T07:12:36.209Z | reconnect         | Room1    | Room       | temperature | centigrade | 16        | [{"name":"resolution","type":"integer","value":"8.00"}]  |
| 1455261156209 | 2016-02-12T07:12:36.209Z | reconnect         | Room1    | Room       | pressure    | pascal     | 16        | [{"name":"resolution","type":"integer","value":"8.00"}]  |
| 1455261157259 | 2016-02-12T07:12:37.259Z | reconnect         | Room1    | Room       | temperature | centigrade | 17        | [{"name":"resolution","type":"integer","value":"8.50"}]  |
| 1455261157259 | 2016-02-12T07:12:37.259Z | reconnect         | Room1    | Room       | pressure    | pascal     | 17        | [{"name":"resolution","type":"integer","value":"8.50"}]  |
| 1455261158302 | 2016-02-12T07:12:38.302Z | reconnect         | Room1    | Room       | temperature | centigrade | 18        | [{"name":"resolution","type":"integer","value":"9.00"}]  |
| 1455261158302 | 2016-02-12T07:12:38.302Z | reconnect         | Room1    | Room       | pressure    | pascal     | 18        | [{"name":"resolution","type":"integer","value":"9.00"}]  |
| 1455261159330 | 2016-02-12T07:12:39.330Z | reconnect         | Room1    | Room       | temperature | centigrade | 19        | [{"name":"resolution","type":"integer","value":"9.50"}]  |
| 1455261159330 | 2016-02-12T07:12:39.330Z | reconnect         | Room1    | Room       | pressure    | pascal     | 19        | [{"name":"resolution","type":"integer","value":"9.50"}]  |
| 1455261160361 | 2016-02-12T07:12:40.361Z | reconnect         | Room1    | Room       | temperature | centigrade | 20        | [{"name":"resolution","type":"integer","value":"10.00"}] |
| 1455261160361 | 2016-02-12T07:12:40.361Z | reconnect         | Room1    | Room       | pressure    | pascal     | 20        | [{"name":"resolution","type":"integer","value":"10.00"}] |
| 1455261161394 | 2016-02-12T07:12:41.394Z | reconnect         | Room1    | Room       | temperature | centigrade | 21        | [{"name":"resolution","type":"integer","value":"10.50"}] |
| 1455261161394 | 2016-02-12T07:12:41.394Z | reconnect         | Room1    | Room       | pressure    | pascal     | 21        | [{"name":"resolution","type":"integer","value":"10.50"}] |
| 1455261162426 | 2016-02-12T07:12:42.426Z | reconnect         | Room1    | Room       | temperature | centigrade | 22        | [{"name":"resolution","type":"integer","value":"11.00"}] |
| 1455261162426 | 2016-02-12T07:12:42.426Z | reconnect         | Room1    | Room       | pressure    | pascal     | 22        | [{"name":"resolution","type":"integer","value":"11.00"}] |
| 1455261163453 | 2016-02-12T07:12:43.453Z | reconnect         | Room1    | Room       | temperature | centigrade | 23        | [{"name":"resolution","type":"integer","value":"11.50"}] |
| 1455261163453 | 2016-02-12T07:12:43.453Z | reconnect         | Room1    | Room       | pressure    | pascal     | 23        | [{"name":"resolution","type":"integer","value":"11.50"}] |
| 1455261164481 | 2016-02-12T07:12:44.481Z | reconnect         | Room1    | Room       | temperature | centigrade | 24        | [{"name":"resolution","type":"integer","value":"12.00"}] |
| 1455261164481 | 2016-02-12T07:12:44.481Z | reconnect         | Room1    | Room       | pressure    | pascal     | 24        | [{"name":"resolution","type":"integer","value":"12.00"}] |
| 1455261165509 | 2016-02-12T07:12:45.509Z | reconnect         | Room1    | Room       | temperature | centigrade | 25        | [{"name":"resolution","type":"integer","value":"12.50"}] |
| 1455261165509 | 2016-02-12T07:12:45.509Z | reconnect         | Room1    | Room       | pressure    | pascal     | 25        | [{"name":"resolution","type":"integer","value":"12.50"}] |
| 1455261166541 | 2016-02-12T07:12:46.541Z | reconnect         | Room1    | Room       | temperature | centigrade | 26        | [{"name":"resolution","type":"integer","value":"13.00"}] |
| 1455261166541 | 2016-02-12T07:12:46.541Z | reconnect         | Room1    | Room       | pressure    | pascal     | 26        | [{"name":"resolution","type":"integer","value":"13.00"}] |
| 1455261167573 | 2016-02-12T07:12:47.573Z | reconnect         | Room1    | Room       | temperature | centigrade | 27        | [{"name":"resolution","type":"integer","value":"13.50"}] |
| 1455261167573 | 2016-02-12T07:12:47.573Z | reconnect         | Room1    | Room       | pressure    | pascal     | 27        | [{"name":"resolution","type":"integer","value":"13.50"}] |
| 1455261230895 | 2016-02-12T07:13:50.895Z | reconnect         | Room1    | Room       | temperature | centigrade | 88        | [{"name":"resolution","type":"integer","value":"44.00"}] |
| 1455261230895 | 2016-02-12T07:13:50.895Z | reconnect         | Room1    | Room       | pressure    | pascal     | 88        | [{"name":"resolution","type":"integer","value":"44.00"}] |
| 1455261231928 | 2016-02-12T07:13:51.928Z | reconnect         | Room1    | Room       | temperature | centigrade | 89        | [{"name":"resolution","type":"integer","value":"44.50"}] |
| 1455261231928 | 2016-02-12T07:13:51.928Z | reconnect         | Room1    | Room       | pressure    | pascal     | 89        | [{"name":"resolution","type":"integer","value":"44.50"}] |
| 1455261232962 | 2016-02-12T07:13:52.962Z | reconnect         | Room1    | Room       | temperature | centigrade | 90        | [{"name":"resolution","type":"integer","value":"45.00"}] |
| 1455261232962 | 2016-02-12T07:13:52.962Z | reconnect         | Room1    | Room       | pressure    | pascal     | 90        | [{"name":"resolution","type":"integer","value":"45.00"}] |
| 1455261233991 | 2016-02-12T07:13:53.991Z | reconnect         | Room1    | Room       | temperature | centigrade | 91        | [{"name":"resolution","type":"integer","value":"45.50"}] |
| 1455261233991 | 2016-02-12T07:13:53.991Z | reconnect         | Room1    | Room       | pressure    | pascal     | 91        | [{"name":"resolution","type":"integer","value":"45.50"}] |
| 1455261235019 | 2016-02-12T07:13:55.19Z  | reconnect         | Room1    | Room       | temperature | centigrade | 92        | [{"name":"resolution","type":"integer","value":"46.00"}] |
| 1455261235019 | 2016-02-12T07:13:55.19Z  | reconnect         | Room1    | Room       | pressure    | pascal     | 92        | [{"name":"resolution","type":"integer","value":"46.00"}] |
| 1455261236044 | 2016-02-12T07:13:56.44Z  | reconnect         | Room1    | Room       | temperature | centigrade | 93        | [{"name":"resolution","type":"integer","value":"46.50"}] |
| 1455261236044 | 2016-02-12T07:13:56.44Z  | reconnect         | Room1    | Room       | pressure    | pascal     | 93        | [{"name":"resolution","type":"integer","value":"46.50"}] |
| 1455261237072 | 2016-02-12T07:13:57.72Z  | reconnect         | Room1    | Room       | temperature | centigrade | 94        | [{"name":"resolution","type":"integer","value":"47.00"}] |
| 1455261237072 | 2016-02-12T07:13:57.72Z  | reconnect         | Room1    | Room       | pressure    | pascal     | 94        | [{"name":"resolution","type":"integer","value":"47.00"}] |
| 1455261238099 | 2016-02-12T07:13:58.99Z  | reconnect         | Room1    | Room       | temperature | centigrade | 95        | [{"name":"resolution","type":"integer","value":"47.50"}] |
| 1455261238099 | 2016-02-12T07:13:58.99Z  | reconnect         | Room1    | Room       | pressure    | pascal     | 95        | [{"name":"resolution","type":"integer","value":"47.50"}] |
| 1455261239127 | 2016-02-12T07:13:59.127Z | reconnect         | Room1    | Room       | temperature | centigrade | 96        | [{"name":"resolution","type":"integer","value":"48.00"}] |
| 1455261239127 | 2016-02-12T07:13:59.127Z | reconnect         | Room1    | Room       | pressure    | pascal     | 96        | [{"name":"resolution","type":"integer","value":"48.00"}] |
| 1455261240153 | 2016-02-12T07:14:00.153Z | reconnect         | Room1    | Room       | temperature | centigrade | 97        | [{"name":"resolution","type":"integer","value":"48.50"}] |
| 1455261240153 | 2016-02-12T07:14:00.153Z | reconnect         | Room1    | Room       | pressure    | pascal     | 97        | [{"name":"resolution","type":"integer","value":"48.50"}] |
| 1455261241181 | 2016-02-12T07:14:01.181Z | reconnect         | Room1    | Room       | temperature | centigrade | 98        | [{"name":"resolution","type":"integer","value":"49.00"}] |
| 1455261241181 | 2016-02-12T07:14:01.181Z | reconnect         | Room1    | Room       | pressure    | pascal     | 98        | [{"name":"resolution","type":"integer","value":"49.00"}] |
| 1455261242209 | 2016-02-12T07:14:02.209Z | reconnect         | Room1    | Room       | temperature | centigrade | 99        | [{"name":"resolution","type":"integer","value":"49.50"}] |
| 1455261242209 | 2016-02-12T07:14:02.209Z | reconnect         | Room1    | Room       | pressure    | pascal     | 99        | [{"name":"resolution","type":"integer","value":"49.50"}] |
| 1455261243234 | 2016-02-12T07:14:03.234Z | reconnect         | Room1    | Room       | temperature | centigrade | 100       | [{"name":"resolution","type":"integer","value":"50.00"}] |
| 1455261243234 | 2016-02-12T07:14:03.234Z | reconnect         | Room1    | Room       | pressure    | pascal     | 100       | [{"name":"resolution","type":"integer","value":"50.00"}] |
| 1455261244263 | 2016-02-12T07:14:04.263Z | reconnect         | Room1    | Room       | temperature | centigrade | 101       | [{"name":"resolution","type":"integer","value":"50.50"}] |
| 1455261244263 | 2016-02-12T07:14:04.263Z | reconnect         | Room1    | Room       | pressure    | pascal     | 101       | [{"name":"resolution","type":"integer","value":"50.50"}] |
| 1455261245287 | 2016-02-12T07:14:05.287Z | reconnect         | Room1    | Room       | temperature | centigrade | 102       | [{"name":"resolution","type":"integer","value":"51.00"}] |
| 1455261245287 | 2016-02-12T07:14:05.287Z | reconnect         | Room1    | Room       | pressure    | pascal     | 102       | [{"name":"resolution","type":"integer","value":"51.00"}] |
| 1455261246320 | 2016-02-12T07:14:06.320Z | reconnect         | Room1    | Room       | temperature | centigrade | 103       | [{"name":"resolution","type":"integer","value":"51.50"}] |
| 1455261246320 | 2016-02-12T07:14:06.320Z | reconnect         | Room1    | Room       | pressure    | pascal     | 103       | [{"name":"resolution","type":"integer","value":"51.50"}] |
| 1455261247352 | 2016-02-12T07:14:07.352Z | reconnect         | Room1    | Room       | temperature | centigrade | 104       | [{"name":"resolution","type":"integer","value":"52.00"}] |
| 1455261247352 | 2016-02-12T07:14:07.352Z | reconnect         | Room1    | Room       | pressure    | pascal     | 104       | [{"name":"resolution","type":"integer","value":"52.00"}] |
| 1455261248377 | 2016-02-12T07:14:08.377Z | reconnect         | Room1    | Room       | temperature | centigrade | 105       | [{"name":"resolution","type":"integer","value":"52.50"}] |
| 1455261248377 | 2016-02-12T07:14:08.377Z | reconnect         | Room1    | Room       | pressure    | pascal     | 105       | [{"name":"resolution","type":"integer","value":"52.50"}] |
| 1455261249406 | 2016-02-12T07:14:09.406Z | reconnect         | Room1    | Room       | temperature | centigrade | 106       | [{"name":"resolution","type":"integer","value":"53.00"}] |
| 1455261249406 | 2016-02-12T07:14:09.406Z | reconnect         | Room1    | Room       | pressure    | pascal     | 106       | [{"name":"resolution","type":"integer","value":"53.00"}] |
| 1455261250434 | 2016-02-12T07:14:10.434Z | reconnect         | Room1    | Room       | temperature | centigrade | 107       | [{"name":"resolution","type":"integer","value":"53.50"}] |
| 1455261250434 | 2016-02-12T07:14:10.434Z | reconnect         | Room1    | Room       | pressure    | pascal     | 107       | [{"name":"resolution","type":"integer","value":"53.50"}] |
+---------------+--------------------------+-------------------+----------+------------+-------------+------------+-----------+----------------------------------------------------------+
96 rows in set (0.00 sec)
```

As can be seen, Cygnus successfully reconnects.

It must be noticed this test has been done by using the batching mechanism (i.e. a batch size greater than 1).

**Comment by frbattid** [comment]
I've tested the same with batch size 1 and the tests fails. But that's because some kind of error in the 1 by 1 processing of events (i.e. when batch size == 1). Thus, there are two options:

* Use a batch size greater than 1. The usage of batches is recommended, not only as a workaround for this issue, but as a configuration improving the performance of Cygnus.
* Fix the 1 by 1 processing in `release/0.11.0`. Nevertheless, this 1 by 1 processing will be definitely disabled in `release/0.13.0` and always a batch will be processed (even in the case of batches of size 1). Thus, this options seems to be working on a feature that wont be available in future releases of IoTP.

**Comment by frbattid** [comment]
Closing since the chosen option for `release/0.11.0` was using a batch size greater than 1.
